File: python/create_default_scenario_and_run.py
# Create a default scenario attached to the supplied project request ID and run it.
# Provide project request ID as a command line argument.
#
# python create_default_scenario_and_run.py <project_request_id>
#

import json
import urllib.request
import os, sys
from time import time, sleep
import api_helpers

#Get app settings
connection, headers, savePath = api_helpers.get_appsettings()

#Define how frequently to check the project's creation status (seconds)
pollInterval = 10 

inputData = {
	'projectRequestId': int(sys.argv[1]),
	'weatherDataset': 'PRISM',
    'startingSimulationDate': '1981-01-01',
    'endingSimulationDate': '1989-12-31',
    'warmupYears': 2,
    'outputPrintSetting': 'daily',
	'writeSwatEditorDb': 'access',
    'reportData': {
        'formats': [ 'csv', 'netcdf' ],
        'units': 'metric',
        'outputs': {
            'rch': {
                'statistics': [ 'daily_avg' ]
            }
        }
    }
}

#Submit the request to the API and read the response
connection.request('POST', '/builder/scenario/create-and-run', json.dumps(inputData), headers)
postResponse = connection.getresponse()
submissionResult = json.loads(postResponse.read().decode())

#Check the status of the project creation until it's complete
print('Scenario request ID {} submitted'.format(submissionResult['id']))
starttime = time()
spaces = 0
while True:
	scenario, spaces = api_helpers.getStatus(connection, headers, '/builder/scenario/{}'.format(submissionResult['id']), spaces)
	if scenario['status']['progress'] >= 100:
		print()
		break
	sleep(pollInterval - ((time() - starttime) % pollInterval))

error = scenario['status']['errorStackTrace']
if error is not None:
	print('Error stack trace: {}'.format(error))

requestPath = os.path.join(savePath, 'Scenario_{}'.format(submissionResult['id']))
os.makedirs(requestPath, exist_ok=True)
for file in scenario['output']:
	print('Retrieving and saving {} ({})'.format(file['name'], file['format']))
	filename = file['url'].split('/')[-1]
	urllib.request.urlretrieve(file['url'], os.path.join(requestPath, filename))

print('Scenario request ID {} creation complete'.format(submissionResult['id']))

File: r/create_default_scenario_and_run.R
# Create a default scenario attached to the supplied project request ID and run it.
# Provide project request ID as a command line argument.
#
# python create_default_scenario_and_run.R <project_request_id>
#


library(httr2)
library(jsonlite)

source("api_helpers.R")

#Get app settings
Appsettings = getAppsettings('appsettings.json')

connection = Appsettings[['connection']]
headers = Appsettings[['headers']]
savePath = Appsettings[['savePath']]
apiUrl = Appsettings[['apiUrl']]

#Define how frequently to check the project's creation status (seconds)
pollInterval = 10 

input_data <- list(
    projectRequestId = as.integer(commandArgs(trailingOnly = TRUE)[1]),
    weatherDataset = "PRISM",
    startingSimulationDate = "1981-01-01",
    endingSimulationDate = "1985-12-31",
    warmupYears = 2,
    outputPrintSetting = "daily",
    reportData = list(
      formats = list("csv", "netcdf"),
      units = "metric",
      outputs = list(
        rch = list(
          statistics = list("daily_avg")
      )
    )
  )
)

#Submit the request to the API and read the response
connection.request = request(apiUrl) |> 
  req_url_path("/builder/scenario/create-and-run") |>
  req_headers(!!!headers) |>
  req_body_json(input_data)

postResponse = connection.request |> req_perform(verbosity = 0)
submissionResult = postResponse |> resp_body_json()

#Check the status of the project creation until it's complete
print(paste("Project request ID", submissionResult[['id']], "submitted"))
spaces = 0

repeat {
  scenario <- getStatus(headers, submissionResult[['url']], spaces)[['project']]
  if (scenario[['status']][['progress']] >= 100) {
    cat("\n")
    break
  }
  Sys.sleep(pollInterval)
}

error = scenario['status']['errorStackTrace'][[1]]
if (!is.null(error)){
  #print('Error stack trace: {}'.format(error))
  print(paste("Error stack trace:", error))
}

#Save scenario output files to disk
requestPath = file.path(savePath, paste0("Scenario_",submissionResult[['id']]))
dir.create(requestPath, recursive=TRUE)


for (file in scenario[['output']]) {
  cat("Retrieving and saving ", file[['name']], " (", file[['format']], ")\n", sep = "")
  filename <- tail(strsplit(file[['url']], "/")[[1]], 1)
  download.file(file[['url']], destfile = file.path(requestPath, filename), mode = "wb")
}

print(paste("Scenario request ID", submissionResult[['id']], "creation complete"))


